modules/code-editor/src/views/full/FileNavigator.tsx

```diff
@@ -77,14 +83,87 @@ export default class FileNavigator extends React.Component<Props, State> {
     }
   }
 
+  handleContextMenu = (node: ITreeNode<NodeData>, path, e) => {
+    if (!node.nodeData) {
+      return null
+    }
+
+    e.preventDefault()
+    ContextMenu.show(
+      <Menu>
+        <MenuItem icon="edit" text="Rename" onClick={() => this.handleRename(node)} />
+        <MenuItem icon="delete" text="Delete" onClick={() => this.props.onNodeDelete(node.nodeData as EditableFile)} />
+      </Menu>,
+      { left: e.clientX, top: e.clientY }
+    )
+  }
+
+  handleRename = (node: ITreeNode<NodeData>) => {
+    const nodeDomElement = this.treeRef.current.getNodeContentElement(node.id)
+
+    const input = document.createElement('input')
+    input.type = 'text'
+    input.className = 'bp3-input bp3-small'
+    input.value = node.label as string
+
+    const div = document.createElement('div')
+    div.className = nodeDomElement.className
+    div.appendChild(input)
+
+    nodeDomElement.replaceWith(div)
+
+    input.focus()
+    input.select()
+
+    const closeRename = async e => {
+      e.preventDefault()
+      div.replaceWith(nodeDomElement)
+      window.removeEventListener('keydown', keyboardListener)
+      window.removeEventListener('mousedown', mouseListener)
+
+      const file = node.nodeData as EditableFile
+      let newName = input.value as string
+      newName = newName.endsWith('.js') ? newName : newName + '.js'
+
+      if (newName === node.label) {
+        return
+      }
+
+      try {
+        await this.props.onNodeRename(file, newName)
```

Comment:
UX wise, it would make sense to validate the input as the user is typing, so the final name is valid (using the same validation regex as the backend). Here you can type //~~^^ to make an invalid name and it will not be able to reach the backend (because of the slash)

Reply:
I made the validation in the backend in only one place and I recover in the frontend with a toaster that indicates that the name is invalid.